### PR TITLE
chore(cardinal): benchmark the tick additionally write a small benchmarking suite

### PR DESCRIPTION
**Pull request by pyrofolium**
Closes: WORLD-718

this PR adds zero new features to world engine. It is simply a modified copy of existing e2e tests optimized for profiling. Profiling files can are in the profiles folder. A summary of the results of the profiling can be seen on slack and my conversations with scott on the team-infra channel

This stuff is mostly internal. I have additional comments near other places in the code. 

To run the tests run `make e2e-benchmark` profiles will appear in the profiles folder. Edit the docker file to change the flag to see which test to run. 

This PR doesn't have to go up. It's just internal stuff, but I recommend we do for the future case of when we want to benchmark again. Otherwise someone will have to rewrite this stuff. 

## Summary by CodeRabbit

- **New Features**
  - Introduced functionality for creating and manipulating entities within a world simulation.
  - Added test suite for benchmarking the Nakama system.
  - Implemented a chain struct with methods for creating a new chain client and initializing gRPC connections.
  - Introduced game simulation components for benchmarking various game operations.
  - Added containerized environments for running Golang applications and integration tests.

- **Bug Fixes**
  - Ensured transaction storage verification is conditional on specific docker container availability.

- **Documentation**
  - Provided instructions for setting up and executing benchmark tests.

- **Chores**
  - Added a new makefile target `e2e-benchmark` for improved benchmark testing workflows.

### CODE REVIEW

File: internal/nakamabenchmark/testsuite_test.go
Comment (namrapatel) [inline internal/nakamabenchmark/testsuite_test.go:1]: What are we trying to do with the `nakamabenchmark` package? It seems like `/nakamabenchmark/testsuite_test.go` and `/nakamabenchmark/util.go` both have a bunch of commented out code?

File: internal/e2ebenchmark/chain_client.go
Comment (namrapatel) [inline internal/e2ebenchmark/chain_client.go:19]: Looks like chain client and nakama client in `e2ebenchmark` are the same as the clients in `e2e`, could we possibly combine the 4 clients into 2 clients and put them in a sibling package in `/internal`?

File: internal/nakamabenchmark/testsuite_test.go
Comment (pyrofolium) [reply internal/nakamabenchmark/testsuite_test.go:1]: This is a test file for the future. It's copied over from the e2e tests. The reason why it's still here because while you're running the benchmark one can run the e2e tests as well. 

In the e2e tests one can run things like external http clients. This allows for realistic benchmarking under realistic loads. The code to do this will go here in the future and it exists in the form of a "testing" file. 

Right now it doesn't do anything, but it's here for the future. I commented out the code because it can be used to create test code as a reference.

File: internal/e2ebenchmark/chain_client.go
Comment (pyrofolium) [reply internal/e2ebenchmark/chain_client.go:19]: I would say no to this. The code is copied over from e2e tests, but what will go in here and what these utlities will do is NOT e2e tests so even this could be modified. I repeat, this code is for benchmarking, and this client will very likely diverge from it's original e2e purpose. 

If we want to use some shared utility package I would say this is possible. But better to wait for a future PR where we know what we want more clearly. It's always better to make a design decision in the future when we actually have a practical issue to solve and understand rather then making a possibly wrong decision now and then possibly "refactoring" it again in the future.

File: internal/nakamabenchmark/testsuite_test.go
Comment (pyrofolium) [reply internal/nakamabenchmark/testsuite_test.go:1]: @namrapatel if you want me to just delete the files and comments, I can do that. I don't feel particularly strongly about this one.

File: internal/e2ebenchmark/chain_client.go
Comment (pyrofolium) [reply internal/e2ebenchmark/chain_client.go:19]: @namrapatel Same here, if you want me to just take this out. I can. But my rec is to leave it in.

File: internal/e2ebenchmark/tester/game/sys/jsys.go
Comment (namrapatel) [inline internal/e2ebenchmark/tester/game/sys/jsys.go:51]: It would be really nice if we could name the systems after what they do here, or at least just leave a short comment above each one. Much easier to read thru them that way since they are just named after a single letter currently.

**Review by namrapatel — CHANGES_REQUESTED** [review]

File: internal/nakamabenchmark/testsuite_test.go
Comment (namrapatel) [reply internal/nakamabenchmark/testsuite_test.go:1]: I see, I'm personally OK with leaving this commented code here for use in the near future, insofar as there is a ticket open for that work. I say we let Jeremy and/or Tyler chime in on this and make a final decision, I'm in favour of leaving it in and opening a ticket

File: internal/e2ebenchmark/chain_client.go
Comment (namrapatel) [reply internal/e2ebenchmark/chain_client.go:19]: I think if the clients diverge in the future, then we should split them into 4 clients at that point. Currently, and seemingly in the near future, the 2 clients can be shared by `e2e` and `e2ebenchmark` both, so I'd be in favour of having a sibling package in `/internal` called something like `utils` or `clients` that can be shared by `e2e` and `e2ebenchmark`.

Would also be good to get @jerargus and/or @technicallyty to chime in on this one to get some more opinions

File: internal/e2ebenchmark/tester/game/sys/jsys.go
Comment (pyrofolium) [reply internal/e2ebenchmark/tester/game/sys/jsys.go:51]: These are mostly Jeremies requested tests. He called it System A and B in his doc. Yeah I can change this.

File: internal/nakamabenchmark/testsuite_test.go
Comment (pyrofolium) [reply internal/nakamabenchmark/testsuite_test.go:1]: There is no ticket. Because we literally don't know what we need to benchmark in the future. 

I only know that we WILL benchmark something in the future. Commenting the code was a practical action here, I realize it's "ugly" but in the future if you do need to write some clients, the file and references are right here. 

Whatever, I can delete it if you feel your right and I'm wrong.

Comment (namrapatel) [reply internal/nakamabenchmark/testsuite_test.go:1]: Nah, I'm cool with small amounts of commented code as long as we think it's actually good, I see what you mean re: no currently opened ticket. Unless someone else takes issue with this in the next day I think you can resolve this one

Comment (jerargus) [reply internal/nakamabenchmark/testsuite_test.go:1]: I don't agree with having commented out files in our code base (this goes for the [game/query/location.go](https://github.com/Argus-Labs/world-engine/pull/540/files#diff-c62d41ff1949849882121655f821c850fd2324c9fcab3f2e590732fc5c6761d3) file above). 

We can make a couple of triage tickets in Linear referencing the source of code code, titled something like "Consider Query benchmarks" and in the body "Use the query code at internal/e2e/tester/game/query for an example".

Comment (pyrofolium) [reply internal/nakamabenchmark/testsuite_test.go:1]: Are you only referring to query? That's more easily removable. @jerargus

Comment (pyrofolium) [reply internal/nakamabenchmark/testsuite_test.go:1]: @jerargus, removed. This one is less less critical. I feel more strongly about the test files, we can just wait for Tylers vote on those.

Comment (pyrofolium) [reply internal/nakamabenchmark/testsuite_test.go:1]: Also guys, none of this is nearly as bad as cancelling the entire events feature imo and leaving the dead code in there lol.

Comment (jerargus) [reply internal/nakamabenchmark/testsuite_test.go:1]: I think both game/query/location.go and this entire internal/nakamabenchmark directory should be removed.

File: internal/e2ebenchmark/chain_client.go
Comment (pyrofolium) [reply internal/e2ebenchmark/chain_client.go:19]: maybe let's do this in another PR.

File: internal/nakamabenchmark/testsuite_test.go
Comment (pyrofolium) [reply internal/nakamabenchmark/testsuite_test.go:1]: @jerargus How about this. I add general code that will load test benchmark in there. A single test and that will make the code relevant for the e2e benchmark.

**Review by namrapatel — CHANGES_REQUESTED** [review]
Should we also `gitignore` the stuff in `profiles`? Not sure how that works, feel free to resolve this comment if silly

**Comment by pyrofolium** [comment]
> Should we also `gitignore` the stuff in `profiles`? Not sure how that works, feel free to resolve this comment if silly

I put it in there so people can view the actual flame graphs and drill into the results. It was just convenient for me to save it there and distribute it to everyone. I think it's educational for everyone to take a look. I can just upload it to slack and remove this folder all together. And Yeah I can add a gitignore.

One thing we can do though is version control the benchmarks so we can see how the engine improves over time.

File: internal/nakamabenchmark/testsuite_test.go
Comment (pyrofolium) [reply internal/nakamabenchmark/testsuite_test.go:1]: lmk if you guys are all cool with that and I'll do it.

Comment (jerargus) [reply internal/nakamabenchmark/testsuite_test.go:1]: I think you should leave this out and add a triage ticket to Linear to benchmark queries and a separate triage ticket to benchmark Nakama CQL and Events stuff.

Comment (pyrofolium) [reply internal/nakamabenchmark/testsuite_test.go:1]: You're saying do this in a separate PR? @jerargus. I would agree with that. The problem is I'll be blocked on this until way later. I think a quick test here to make the load more realistic in the profile works. 

We can vote on it. @technicallyty what do you think?

Comment (technicallyty) [reply internal/nakamabenchmark/testsuite_test.go:1]: agree with jeremy, would prefer to just throw the commented out code into a triage ticket. the difference between this and the events situation is, we know what to do and have a ticket for addressing the events. there is a plan, and people are assigned and know what to do when the server stuff is completed. here, the dead code is just here for a rainy day.

i vote to toss it into triage as jeremy suggested above

Comment (pyrofolium) [reply internal/nakamabenchmark/testsuite_test.go:1]: @namrapatel @jerargus @technicallyty 

I talked to tyler in person, He seemed convinced, it seems that from that conversation he and most people aren't fully grasping what I'm doing with the profiling PR here. It's not possible to fully understand what this is if you weren't following my conversations with Scott on slack which is fine, I made the mistake of assuming people were but there's too much going on here. 

I'll explain it more in depth in stand up tomorrow. And I'll show you the results of it, it's actually very visual and pretty cool.

File: internal/e2ebenchmark/chain_client.go
Comment (namrapatel) [reply internal/e2ebenchmark/chain_client.go:19]: Made ticket [here](https://linear.app/arguslabs/issue/WORLD-747/merge-client-code-in-e2e-and-e2ebenchmark), resolving

**Review by namrapatel — APPROVED** [review]
I think this is ready with the 2 changes I am aware were outstanding:

- System names
- Delete unused code and use the parts of the nakamabenchmark package that remain

File: internal/nakamabenchmark/testsuite_test.go
Comment (pyrofolium) [reply internal/nakamabenchmark/testsuite_test.go:1]: As discussed I added the http client to the profiling. You can see the load it adds to the profile here:

Screenshot 2024-01-24 at 12 30 47 PM

Just want to reiterate. The nature of the flame graph makes it pointless to test things by themselves. So if I test just redis, or just parsing, or just the http server I get no context. The flamegraph only gives you benchmarking info about things Relative to OTHER things. So you actually have to benchmark more than one thing to even get this style of profiling to work.

File: internal/e2ebenchmark/chain_client.go
Comment (pyrofolium) [reply internal/e2ebenchmark/chain_client.go:19]: @namrapatel created a ticket in the backlog, I will get to it when needed.

**Comment by pyrofolium** [comment]
> I think this is ready with the 2 changes I am aware were outstanding:
> 
> * System names
> * Delete unused code and use the parts of the nakamabenchmark package that remain

These are done.

**Pull request merged**
